src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests.cs

```diff
@@ -474,5 +476,124 @@ private ResourceWrapper CreateMockResourceWrapper(ResourceElement resource, bool
                 null,
                 0);
         }
+
+        [Fact]
+        public async Task GivenUpsertRequestWithPostHttpVerb_WhenUserHasCreatePermission_ThenShouldSucceed()
+        {
+            var resource = Samples.GetDefaultObservation();
+            var contextAccessor = Substitute.For<FhirRequestContextAccessor>();
+            contextAccessor.RequestContext = new FhirRequestContext("POST", "http://localhost", "http://localhost", "id", new Dictionary<string, StringValues>(), new Dictionary<string, StringValues>());
+            var bundleContext = new BundleResourceContext(BundleProcessingLogic.Sequential, HTTPVerb.POST, Guid.NewGuid());
+
+            _authorizationService.CheckAccess(DataActions.Create | DataActions.Write, Arg.Any<CancellationToken>()).Returns(DataActions.Create);
+            _fhirDataStore.UpsertAsync(Arg.Any<ResourceWrapperOperation>(), Arg.Any<CancellationToken>()).Returns(x => new UpsertOutcome(x.ArgAt<ResourceWrapperOperation>(0).Wrapper, SaveOutcomeType.Created));
+
+            var handler = new UpsertResourceHandler(_fhirDataStore, new Lazy<IConformanceProvider>(() => _conformanceProvider), _resourceWrapperFactory, _resourceIdProvider, new ResourceReferenceResolver(_searchService, new TestQueryStringParser(), Substitute.For<ILogger<ResourceReferenceResolver>>()), contextAccessor, _authorizationService, ModelInfoProvider.Instance);
+            var request = new UpsertResourceRequest(resource, bundleContext);
+
+            var result = await handler.Handle(request, CancellationToken.None);
+            Assert.NotNull(result);
+        }
+
+        [Fact]
+        public async Task GivenUpsertRequestWithPostHttpVerb_WhenUserLacksCreatePermission_ThenShouldThrowUnauthorizedException()
```

Comment:
Should we add - Post with Update permission should fail?

Reply:
Added.